kernel/posix/pthread_semaphore.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+
+typedef struct pthread_sem {
+	struct k_sem *sem;
+} sem_t;
+
+int sem_destroy(sem_t *semaphore)
+{
+	k_sem_reset(semaphore);
+	return 0;
+}
+
+int sem_getvalue(sem_t *semaphore, int *value)
+{
+	int count;
+
+	count = k_sem_count_get(semaphore);
+	*value = count;
+
+	return 0;
+}
+
+int sem_init(sem_t *semaphore, int pshared, unsigned value)
+{
+	if (semaphore->limit = 0){
+		/* limit can not be zero*/
+		return EINVAL;
+	}
+	ARG_UNUSED(pshared);
+	k_sem_init(semaphore, value, semaphore->limit);
+
+	return 0;
+}
+
+int sem_post(sem_t *semaphore)
+{
+	k_sem_give(semaphore);
+	return 0;
+}
+
+int sem_timedwait(sem_t *semaphore, const struct timespec *abstime)
+{
+	u32_t 		timeout;
+	long		msecs = 0;
+	int 		retc = 0;
+	s64_t		current_ms;
+	int		abstime_ms;
+	bool		retval;
+
+	if ((abstime->tv_sec < 0) || (1000000000 <= abstime->tv_nsec)) {
+		return -1;
+	}
+
+	current_ms = k_uptime_get();
+	abstime_ms = _ts_to_ms(abstime);
+	msecs = abstime_ms - current_ms;
+
+	if (abstime_ms <= current_ms) {
+		timeout = 0;
+	} else {
+		timeout = _ms_to_ticks(msecs);
+	}
+
+	retc = k_sem_take(semaphore, timeout);
```

Comment:
`k_sem_take` takes a `k_sem` as the first parameter.  This won't work: not only it'll generate a warning, the first thing at the `semaphore` address is a pointer to a `k_sem` structure, not the structure itself.

Reply:
yeah just changed sem_t to typedef struct k_sem sem_t